robolectric/robolectric
# [Morgan] Implement monitoring foreground app change using a system API (push model).

[Morgan] Implement monitoring foreground app change using a system API (push model).

We've got 2 approaches to implement this:
 - [ProcessObserver](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/IProcessObserver.aidl)
 - [UidImportanceListener](https://android.googlesource.com/platform/frameworks/base.git/+/master/core/java/android/app/ActivityManager.java#3324)

However *ProcessObserver* is a hidden API that is not allowed to be used by anyone. This is by google policy. There might be no enforcements in other repos such as android repo, but google3 is explicitly forbidden to use hidden API per [guidelines](https://g3doc.corp.google.com/company/teams/androidx/api_guidelines.md?cl=head#restricted-api).

Thus in this CL I'm implementing it using the second approach, which is a System API that Katniss is allowed to use.

Sherlog: http://sherlog/_dyqFrVEWiT

BEGIN_PUBLIC  
Add addOnUidImportanceListener and removeOnUidImportanceListener to ShadowActivityManager.  
END_PUBLIC  

QA: On Sabrina
1. Turn off AgsaTv__state_sync_enabled
2. Set AgsaTv__state_sync_state_sync_foreground_app_interval_sec to 0
3. Turn on AgsaTv__state_sync_enabled
4. Run $adb logcat -s ForegroundAppMonitor:D StateSyncer:D ClientInputTransmitter:D
5. Open an app (Youtube, Netflix, etc.)
6. Verify that there is log printed in logcat: "ForegroundAppMonitor: #getEvents(). Getting foreground app events." and "Importance of uid: blabla (app package name) changed to blabla. Fetching foreground app".
7. Verify in sherlog that there is a assistant voiceless request: tv.STATE_UPDATE entry
